src/hotspot/os/windows/os_windows.cpp

```diff
@@ -3982,14 +3982,201 @@ bool   os::win32::_is_windows_server         = false;
+    return; // nothing to do if already the version has already been set
+  }
+
+  LPWSTR kernel32_path = NULL;
+  LPWSTR version_info = NULL;
+  DWORD version_size = 0;
+
+  // Get the full path to \Windows\System32\kernel32.dll and use that for
+  // determining what version of Windows we're running on.
+  UINT buffer_size = GetSystemDirectoryW(NULL, 0);
+  if (buffer_size == 0) {
+    warning("GetSystemDirectoryW() failed: GetLastError->%ld.", GetLastError());
+    return;
+  }
+
+  UINT filename_chars = (UINT)strlen("\\kernel32.dll");
+
+  // buffer_size includes the terminating null character
+  UINT size = (buffer_size + filename_chars) * sizeof(WCHAR);
+  kernel32_path = (LPWSTR)os::malloc(size, mtInternal);
+  if (kernel32_path == NULL) {
+      warning("os::malloc() failed to allocate %ld bytes for the kernel32.dll path", size);
+      return;
+  }
+
+  UINT non_null_chars_written = GetSystemDirectoryW(kernel32_path, size);
+  if (non_null_chars_written == 0) {
+    warning("GetSystemDirectoryW() failed: GetLastError->%ld.", GetLastError());
+    goto free_mem;
+  }
+
+  wcsncat(kernel32_path, L"\\kernel32.dll", filename_chars);
+
+  version_size = GetFileVersionInfoSizeW(kernel32_path, NULL);
+  if (version_size == 0) {
+    warning("GetFileVersionInfoSizeW() failed: GetLastError->%ld.", GetLastError());
+    goto free_mem;
+  }
+
+  version_info = (LPWSTR)os::malloc(version_size, mtInternal);
+  if (version_info == NULL) {
+    warning("os::malloc() failed to allocate %ld bytes for GetFileVersionInfoW buffer", version_size);
+    goto free_mem;
+  }
+
+  if (!GetFileVersionInfoW(kernel32_path, 0, version_size, version_info)) {
+    warning("GetFileVersionInfoW() failed: GetLastError->%ld.", GetLastError());
+    goto free_mem;
+  }
+
+  VS_FIXEDFILEINFO *file_info;
+  if (!VerQueryValueW(version_info, L"\\", (LPVOID*)&file_info, &size)) {
+    warning("VerQueryValueW() failed. Cannot determine Windows version.");
+    goto free_mem;
+  }
+
+  _major_version = HIWORD(file_info->dwProductVersionMS);
+  _minor_version = LOWORD(file_info->dwProductVersionMS);
+  _build_number  = HIWORD(file_info->dwProductVersionLS);
+  _build_minor   = LOWORD(file_info->dwProductVersionLS);
+
+free_mem:
+  os::free(version_info);
+  os::free(kernel32_path);
+}
+
+bool os::win32::is_windows_11_or_greater() {
+  initialize_windows_version();
+
+  // Windows 11 starts at build 22000 (Version 21H2) as per
+  // https://learn.microsoft.com/en-us/windows/release-health/windows11-release-information
+  return (_major_version >= 10 && _build_number >= 22000 && !IsWindowsServer());
+}
+
+bool os::win32::is_windows_server_2022_or_greater() {
+  initialize_windows_version();
+
+  // Windows Server 2022 starts at build 20348.169 as per
+  // https://learn.microsoft.com/en-us/windows/release-health/release-information
+  return (_major_version >= 10 && _build_number >= 20348 && IsWindowsServer());
+}
+
+int os::win32::bit_count(ULONG64 argument) {
```

Comment:
Is there no existing code in HotSpot that does something similar?

Reply:
cc @JohnTortugo

Reply:
Ah, I think population_count is what I needed. I have removed this definition.